File: app/models/fetch_status.rb
class FetchStatus < ApplicationRecord
  belongs_to :fetchable, polymorphic: true, touch: true
  include ActionView::Helpers::NumberHelper

  def percentage_fetched
    number_with_precision(bytes_transferred.to_d / bytes_total.to_d * 100.0, precision: 2) if status == 'LOADING' && bytes_transferred.present?
  end
end

File: app/javascript/app/modules/fetchStatus.ts
interface Success {
  status: "SUCCESS";
  url: string;
  bytesTotal: number;
}

interface Loading {
  status: "LOADING";
  bytesTransferred?: number;
  bytesTotal: number;
  percentageFetched: string;
}

interface Failure {
  status: "FAILURE";
  errorReason: string;
}

interface NotAsked {
  status: "NOT_ASKED";
}

export type FetchStatus = Success | Loading | Failure | NotAsked;
